hedera-node/src/main/java/com/hedera/services/ledger/TransferLogic.java

```diff
@@ -0,0 +1,130 @@
+    private static final List<AccountProperty> TOKEN_TRANSFER_SIDE_EFFECTS =
+            List.of(TOKENS, NUM_NFTS_OWNED, ALREADY_USED_AUTOMATIC_ASSOCIATIONS);
+    private final TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accountsLedger;
+    private final TransactionalLedger<NftId, NftProperty, MerkleUniqueToken> nftsLedger;
+    private final TransactionalLedger<Pair<AccountID, TokenID>, TokenRelProperty, MerkleTokenRelStatus> tokenRelsLedger;
+    private final SideEffectsTracker sideEffectsTracker;
+    private final TokenStore tokenStore;
+    private final MerkleAccountScopedCheck scopedCheck;
+    private final UniqTokenViewsManager tokenViewsManager;
+
+    @Inject
+    public TransferLogic(TransactionalLedger<AccountID, AccountProperty, MerkleAccount> accountsLedger,
+                         TransactionalLedger<NftId, NftProperty, MerkleUniqueToken> nftsLedger,
+                         TransactionalLedger<Pair<AccountID, TokenID>, TokenRelProperty, MerkleTokenRelStatus> tokenRelsLedger,
+                         TokenStore tokenStore,
+                         SideEffectsTracker sideEffectsTracker,
+                         UniqTokenViewsManager tokenViewsManager,
+                         GlobalDynamicProperties dynamicProperties,
+                         OptionValidator validator) {
+        this.accountsLedger = accountsLedger;
+        this.nftsLedger = nftsLedger;
+        this.tokenRelsLedger = tokenRelsLedger;
+        this.sideEffectsTracker = sideEffectsTracker;
+        this.tokenStore = tokenStore;
+        this.tokenViewsManager = tokenViewsManager;
+
+        scopedCheck = new MerkleAccountScopedCheck(dynamicProperties, validator);
+    }
+
+    public void transfer(List<BalanceChange> changes) {
+        var validity = OK;
+        for (var change : changes) {
+            if (change.isForHbar()) {
+                validity = accountsLedger.validate(change.accountId(), scopedCheck.setBalanceChange(change));
+            } else {
+                validity = tokenStore.tryTokenChange(change);
+            }
+            if (validity != OK) {
+                break;
+            }
+        }
+
+        if (validity == OK) {
+            adjustHbarUnchecked(changes);
+        } else {
+            dropPendingTokenChanges();
+            throw new InvalidTransactionException(validity);
+        }
+    }
+
+    private void adjustHbarUnchecked(List<BalanceChange> changes) {
+        for (var change : changes) {
+            if (change.isForHbar()) {
+                final var accountId = change.accountId();
+                final var newBalance = change.getNewBalance();
+                accountsLedger.set(accountId, BALANCE, newBalance);
+                sideEffectsTracker.trackHbarChange(accountId, change.units());
+            }
+        }
+    }
+
+    public void dropPendingTokenChanges() {
+        // TODO: Should we add rollback for tokensLedger (which comes from tokenStore)
```

Comment:
No, we do not need to roll it back since we will not be using it other than for querying tokens